r-data-manipulation.Rmd

```diff
@@ -0,0 +1,1111 @@
+Notice the datasets both have a column called `nta`, and this column contains the NTA code. {TODO: either point out this isn't true in full size data, or make it true by changing data.} This provides a way to match rows in `dogs` to the rows in `neighborhoods`: for each row of `dogs`, look up the value of `nta` in `neighborhoods` and return the values of `population` and `nta_name`.  This might result in a new dataset that looks like this: 
+```{r, echo = FALSE, message = FALSE}
+dogs %>% left_join(neighborhoods)
+```
+Both "PRECIOUS" and "HUNTER" live in Hammels-Arverne-Edgemere with a population of 36,885 and "GUCCI" lives in South Jamaica with a population of 38,894.  Notice that "TOBY" has missing values for both `population` and `nta_name` because his value of `nta`, `"QN55"`, wasn't in this small `neighborhoods` data.  
+
+This join can be performed with the `left_join()` function in dplyr:
+```{r}
+dogs %>% left_join(neighborhoods)
+```
+It's called a `left_join()` because we will get (at least) one row back for every row in the dataset on the left, in this case, `dogs`.  I like to think about `left_join()` working just as we described above: for each row in the dataset on the left, look up the value in the dataset on the right. 
+
+Notice that `left_join()` gave us the message `Joining, by = "nta"`.  The join functions in dplyr will use any columns with the same name in both datasets to do the join.  If the datasets have columns for the same quantity but with different names, for instance, if `neighborhoods` had a column called `nta_code` instead of `nta`:
+```{r}
+neighborhoods_2 <- neighborhoods %>% 
+  rename(nta_code = nta)
+neighborhoods_2 
+```
+
+Then attempting the join will result in an error - `data sources have no common variables`:
+```{r, error = TRUE}
+dogs %>% left_join(neighborhoods_2)
+```
+You can use the `by` argument to the join function to specify which columns are to be used to do the join:
+```{r}
+dogs %>% 
+  left_join(neighborhoods_2, by = c("nta" = "nta_code"))
+```
+
+### Exercise
+
+Try to predict what this code will return, then run it and check your prediction:
+```{r, results = "hide", message = FALSE}
+neighborhoods %>% left_join(dogs)
+```
+
+### Exercise: Dog licenses per capita
+
+* Use `left_join()` to join `dogs_by_nta` with `nta_pop`. 
+
+* `filter()` the result to just use population from `year` 2010.
+
+* Add a `mutate()` step to find the number of dog licenses per person in each NTA.
+
+* Which neighborhood has the highest dog licenses per person based on the 2010 population?
+
+```{r, include = FALSE}
+dogs_by_nta %>% 
+  left_join(nta_pop, by = c("neighborhood_tabulation_area" = "nta_code")) %>% 
+  mutate(licenses_per_person = n_licences/population) %>% 
+  filter(year == 2010) %>% 
+  arrange(desc(licenses_per_person)) %>% 
+  select(nta_name, licenses_per_person)
+# Stuyvesant Town-Cooper Village averages 0.045 dogs per person
+# Or ~45 dogs per 1000 people
+```
+
+### Other join functions
```

Comment:
How important is it to describe these other joins? I personally don't use many other than `left_join` and `full_join` and often struggle with the logic of the other joins. Would it be better to simply say "there are other joins out there, check out the cheatsheet/linked tutorial if you want to learn more, but we won't cover them more here". Thoughts?

Reply:
I almost only use `left_join()`.  So, probably not very important...